File: Python3/symmetric_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root: 
            return True
        else: 
            return self.isMirror(root.left, root.right)
        
    def isMirror(self, p, q):
        if not p and not q: 
            return True
        elif not p or not q:
            return False
        elif p.val != q.val:
            return False
        else:
            return self.isMirror(p.left, q.right) and self.isMirror(p.right, q.left)

File: C++/symmetric_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(root == NULL) return true;
        else return isMirror(root->left, root->right);
    }
    
    bool isMirror(TreeNode* p, TreeNode* q) {
        if(p == NULL and q == NULL) return true;
        else if(p == NULL or q == NULL) return false;
        else if(p->val != q->val) return false;
        else return isMirror(p->left, q->right) and isMirror(p->right, q->left);
    }
};
